crates/rover-client/src/operations/persisted_queries/publish/types.rs

```diff
@@ -107,6 +118,71 @@ impl From<PersistedQueriesPublishInput> for QueryVariables {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RelayPersistedQueryManifest {
+    #[serde(flatten)]
+    operations: HashMap<String, String>,
+}
+
+impl TryFrom<RelayPersistedQueryManifest> for ApolloPersistedQueryManifest {
+    type Error = RoverClientError;
+
+    fn try_from(relay_manifest: RelayPersistedQueryManifest) -> Result<Self, Self::Error> {
+        let mut compiler = ApolloCompiler::new();
+
+        let mut diagnostics = Vec::new();
+        let mut operations = Vec::new();
+        let mut id_mapper = HashMap::new();
+        for (id, body) in &relay_manifest.operations {
+            let file_id = compiler.add_document(body, id);
+            id_mapper.insert(file_id, (id, body));
+        }
+        for operation in compiler.db.all_operations().iter() {
```

Comment:
While I recognize that this is just how apollo-compiler works, I honestly feel like this approach feels hard for me to reason about.  If I understand correctly what's happening is roughly:

- Create a Compiler
- Add each document to the Compiler separately, carefully tracking its FILE ID
- Ask the Compiler to iterate over all operations (but — this is a complex procedure here that involves dicing up the documents into separate operations?)
- Get back the file ID and do some sort of validation over it
- Pull out the operation name and do some logic
- ... honestly I cannot figure out where, if anywhere, "document must have at most one operation" is implemented?

I don't understand why *anything* needs to be shared across multiple document parsing. Doing that makes this a lot harder for me to understand.

This would seem a lot simpler:
- For each document:
  - Parse it in the simplest possible way (probably apollo-parser not apollo-compiler), with no state shared from the previous document.
  - Iterate over the top level definitions:
    - If you see an anonymous operation, that's an error
    - If you see a named operation, remember it, but if you've already seen one, that's an error
    - If you get to the end and you haven't remembered an operation, that's an error.
    - Return the name and type of the operation you remembered.

I'm not really convinced that the apollo-compiler multi-document model helps us here rather than making things more complex and harder to understand.